### PR TITLE
Sync: Migrate 'jetpack_sync_before_send*' actions for Sync queue to 'jetpack_sync_before_enqueue' instead

### LINKED ISSUE #5491: Cart contents cleared when logging in at checkout w/ WCPay and JP active
### Describe the bug
When both WooCommerce Payments and Jetpack are activated, carts created as a guest are cleared when logging in on the checkout page.

Deactivating either plugin fixes the issue.

### To Reproduce

1. Install and connect Jetpack.
2. Install WCPay and start the onboarding. An account doesn't need to finish onboarding, it just needs to connect with Jetpack.
3. Allow customers to log in at checkout under WooCommerce > Settings > Accounts & Privacy.
4. Visit the store while logged out and add some products to your cart
5. Go to checkout, log in, cart is cleared.
6. Deactivate WCPay or JP and clear your carts.
7. Make a new guest cart, log in at checkout, cart contents persist after logging in.

### Actual behavior
Logging in at checkout clears the cart and both the `woocommerce_cart_hash` and `woocommerce_items_in_cart` cookies.

### Expected behavior
I expected the cart contents to persist after logging in at checkout.

### Additional context

**System Status**

<details>

```
### WordPress Environment ###

WordPress address (URL): https://welcome-seal.jurassic.ninja
Site address (URL): https://welcome-seal.jurassic.ninja
WC Version: 7.3.0
REST API Version: ✔ 7.3.0
WC Blocks Version: ✔ 9.1.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.55 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.33
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.3.0
WC Database Prefix: wp_
Total Database Size: 4.87MB
Database Data Size: 3.38MB
Database Index Size: 1.49MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.48MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 3
product: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (6) ###

Companion Plugin: by Osk – 1.29
Jetpack: by Automattic – 11.7.1
WooCommerce Payments Dev Tools: by Automattic –
WooCommerce Payments: by Automattic – 5.3.0
WooCommerce: by Automattic – 7.3.0
WP Rollback: by GiveWP.com – 1.7.3

### Inactive Plugins (3) ###

Akismet Anti-Spam: by Automattic – 5.0.1
Hello Dolly: by Matt Mullenweg – 1.7.2
WooCommerce Shipping & Tax: by WooCommerce – 2.2.1

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
Order datastore: WC_Order_Data_Store_CPT

### WC Pages ###

Shop base: Automattic/jetpack#5 - /shop/
Cart: Automattic/jetpack#6 - /cart/
Checkout: Automattic/jetpack#7 - /checkout/
My account: Automattic/jetpack#8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Three
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Subscriptions Live URL: https://welcome-seal.jurassic.ninja
Subscription Statuses: –
WooCommerce Account Connected: ❌ No

### Store Setup ###

Country / State: United States (US) — California

### Payment Gateway Support ###


### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2023-02-01 00:23:38 +00:00
Options: ✔
Notes: 39
Onboarding: -

### WooCommerce Payments ###

Version: 5.3.0
Connected to WPCOM: Yes
Blog ID: 215089087
Account ID: acct_1MW7isFdW2Ksbbdy

### Action Scheduler ###

Complete: 9
Oldest: 2023-01-31 00:26:39 +0000
Newest: 2023-01-31 00:38:38 +0000

Pending: 2
Oldest: 2023-01-31 01:28:12 +0000
Newest: 2023-02-01 00:24:52 +0000


### Status report information ###

Generated at: 2023-01-31 01:27:48 +00:00

```

</details>

Found in 5913108-zen
This may be related to the enhancement request here: https://github.com/Automattic/wp-super-cache/issues/875
This is also likely the root issue in this forum thread and potentially others: https://wordpress.org/support/topic/cart-clears-when-customer-logins/

### PR DESCRIPTION
**Pull request by fgiannar**
## Proposed changes:
- Migrates `jetpack_sync_before_send_*` actions for the Sync queue to `jetpack_sync_before_enqueue_*` instead.
- ~Updates the Dedicated Sync init hook priority to `0` (was `90`)~ Will do in a follow up PR 

The idea behind this change is to make the Dedicated Sync requests as fast as possible. When a Dedicated Sync request is triggered (Sync queue is unblocked and non empty, aka we need to send those actions to WPCOM), we fetch the Sync queue and send the corresponding actions to WPCOM. Ideally we want this done as fast as possible, thus independent from any other plugin `init` hooks so it can execute and exit early.
However, the current logic behind `jetpack_sync_before_send_*` actions prevented us from doing so. An example is the Woo regression we introduced when we attempted to only update Dedicated Sync's `init` hook priority in the past. 
By only updating the `init` hook priority we ended up with Cache site posts (orders) with `jps_non-reg` statuses and missing fields. See p1HpG7-nxq-p2
This happened because we were essentially checking/updating those queued posts on `jetpack_sync_before_send_jetpack_sync_action` but the corresponding post type (`orders`) was not registered yet when the Dedicated Sync hook priority was updated to `0`.
Therefore this PR moves the corresponding actions that would modify content before sending to before enqueueing instead.

**Pros:**
- More accurate sending of data, since the synced items might have changed between enqueuing and sending
- ~Faster sending of Sync actions to WPCOMFaster sending of Sync actions to WPCOM~ Will do in a follow up PR (related to Dedicated Sync `init` hook priority)
- ~Fixes [Cart contents cleared when logging in at checkout w/ WCPay and JP active 
](https://github.com/Automattic/woocommerce-payments/issues/5491)~ Will do in a follow up PR (related to Dedicated Sync `init` hook priority)

**Cons:**
- Since we are expanding some actions before adding them to the queue, their corresponding size in the DB is going to grow

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)? **N/A**

## Jetpack product discussion

p9dueE-7zI-p2

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:
- Ensure Sync actions are properly synced for the affected modules: Posts, Comments, Users, Updates, Plugins
- Ensure checksums are still valid for the affected modules: Posts, Comments, Users, Updates, Plugins
- Test with WooCommerce and ensure Orders are successfully synced

_A few specific cases TBA / See unit tests_

### CODE REVIEW

File: projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-posts.php [hunk 1]
```diff
@@ -854,6 +879,8 @@ public function test_remove_contact_form_shortcode_from_filtered_content() {
 	}
 
 	public function test_remove_likes_from_filtered_content() {
+		$this->markTestSkipped( 'Temp skipping test' );
```
Comment (fgiannar) [inline projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-posts.php]: I had to mark a few tests as skipped for now, we need to fix those

Comment (fgiannar) [reply projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-posts.php]: Fixed in https://github.com/Automattic/jetpack/pull/33384/commits/c12fe34362bfdcfe47712bbe1a8d00c2619f3ed5

**Review by mdbitz — COMMENTED** [review]
Hi @fgiannar 👋 

I've looked through opengrok across all projects to see if the before_send hooks are being utilized in any projects and the good news is they are not.

I personally think before_enqueue does make sense over before_send as if we expand data at send there is the possibility it will have changed. In testing we have clear actions that are being modified and we should be able to test inserting comment|post and updating plugins and as long as the data is going through as expected I think the risk in this change is addressed.

For roll-out I would like to see the custom table fully released first as this ensures any growth in queue storage doesn't impact the options table. If this was a .com change I would not change the add_dedicated_sync_sender_init from 90 to 0 but as this is part of the Jetpack release it kind of makes sense it happens all at once. Could we leverage Atomic deploys to split the change and verify  before_enqueue changes separately then the init change?

All of this is to say from my view this all makes sense.

**Comment by fgiannar** [comment]
> I've looked through opengrok across all projects to see if the before_send hooks are being utilized in any projects and the good news is they are not.

Thanks, Matthew, really appreciate it! That's also my take :)

> I personally think before_enqueue does make sense over before_send as if we expand data at send there is the possibility it will have changed. In testing we have clear actions that are being modified and we should be able to test inserting comment|post and updating plugins and as long as the data is going through as expected I think the risk in this change is addressed.

💯 

> For roll-out I would like to see the custom table fully released first as this ensures any growth in queue storage doesn't impact the options table. If this was a .com change I would not change the add_dedicated_sync_sender_init from 90 to 0 but as this is part of the Jetpack release it kind of makes sense it happens all at once. Could we leverage Atomic deploys to split the change and verify before_enqueue changes separately then the init change?

Yes, I also believe that changing the add_dedicated_sync_sender_init from 90 to 0 will positively affect the queue lag, especially in sites with lots of plugins performing various actions on init.
I'm fine with splitting the change, this PR is mainly a POC that will enable us to test the suggested approach at once. I'd like us to fully test it first and then I can move modifying the hook priority in a separate PR.

> All of this is to say from my view this all makes sense.

Again, many thanks for taking the time to review this change!

**Review by bisko — APPROVED** [review]
LGTM! 

Tests well and I am excited for the future!

**Comment by naman03malhotra** [comment]
Thanks, folks for working on it. Which version will this PR be shipped with? Asking in the context of https://github.com/Automattic/woocommerce-payments/issues/5491. Shall we update jetpack-related dependencies?

**Comment by fgiannar** [comment]
> Thanks, folks for working on it. Which version will this PR be shipped with? Asking in the context of [Automattic/woocommerce-payments#5491](https://github.com/Automattic/woocommerce-payments/issues/5491). Shall we update jetpack-related dependencies?

Hi @naman03malhotra ! The[ actual fix PR ](https://github.com/Automattic/jetpack/pull/33594) is still not merged as we want to first validate this one on the next AT release and then ship the fix. 
Assuming all goes well the fix should be included in Sync package version `1.58.0`.
I do plan to P2 accordingly so we can test and update the Sync package in WooCommerce Payments as needed :)

**Pull request merged**